### PR TITLE
confdgnmi - sample interval

### PR DESCRIPTION
**Pull request by micnovak**

### CODE REVIEW

File: src/confd_gnmi_adapter.py [hunk 1]
```diff
@@ -156,13 +159,18 @@ def is_poll(self):
...
             Return True if subscription is of such type that we should monitor
             changes.
             :return:
             """
-            return self.subscription_list.mode == gnmi_pb2.SubscriptionList.STREAM
+            is_monitor = False
+            if self.subscription_list.mode == gnmi_pb2.SubscriptionList.STREAM:
+                for s in  self.subscription_list.subscription:
+                    if s.mode == gnmi_pb2.SubscriptionMode.ON_CHANGE:
+                        is_monitor = True
```
Comment (jomiklos) [inline src/confd_gnmi_adapter.py:172]: no need  to iterate all subs after first one hits condition?
just `return True` here, and `return False` at the end of fn?
(no need for local variable `is_monitor`)

File: src/confd_gnmi_adapter.py [hunk 2]
```diff
@@ -279,24 +281,59 @@ def read(self):
...
             event = None
             first_sample = True
+            next_sample_interval = None
+            sample_subscriptions = self.subscription_list.subscription
+
+            def get_next_sample_interval_and_subscriptions(first_sample_time: int):
+                interval = None
+                subscriptions = []
+                for s in self.subscription_list.subscription:
+                    if s.mode == gnmi_pb2.SubscriptionMode.SAMPLE:
```
Comment (jomiklos) [inline src/confd_gnmi_adapter.py:291]: wasteful indent covering whole for loop body -> i'd use `if s.mode != SAMPLE: continue` and save whole body indent level...

File: src/confd_gnmi_client.py [hunk 3]
```diff
@@ -72,13 +72,20 @@ def get_capabilities(self) -> gnmi_pb2.CapabilityResponse:
 
     @staticmethod
     def make_subscription_list(prefix, paths, mode, encoding,
-                               stream_mode=gnmi_pb2.SubscriptionMode.ON_CHANGE):
+                               stream_mode=gnmi_pb2.SubscriptionMode.ON_CHANGE,
+                               sample_interval: int = None):
```
Comment (jomiklos) [inline src/confd_gnmi_client.py:76]: unclear what is the interval unit/resolution -> maybe add `_ms/_s/whatever-is-valid`?

File: src/confd_gnmi_adapter.py [hunk 4]
```diff
@@ -279,24 +281,59 @@ def read(self):
                 assert self.read_queue is not None
             event = None
             first_sample = True
+            next_sample_interval = None
+            sample_subscriptions = self.subscription_list.subscription
+
+            def get_next_sample_interval_and_subscriptions(first_sample_time: int):
+                interval = None
+                subscriptions = []
```
Comment (martin-volf) [inline src/confd_gnmi_adapter.py:289]: This is a function nested in a method of a class nested in a class. I think you are overdoing it.

File: src/confd_gnmi_adapter.py [hunk 5]
```diff
@@ -311,8 +348,14 @@ def read(self):
                     log.warning("**** event=%s not processed ! ****", event)
                     # TODO error
                     break
-                log.debug("Waiting for event")
-                event = self.read_queue.get()
+                log.debug("Waiting for event next_sample_interval=%s",
+                         next_sample_interval)
+                try:
+                    event = self.read_queue.get(timeout=next_sample_interval)
```
Comment (martin-volf) [inline src/confd_gnmi_adapter.py:354]: You saved a thread here, but I don't agree it's worth the price. I believe the code would be quite a lot cleaner with a timer thread injecting events into `read_queue`. (Yes, simple Enum events would not be enough any longer - even then.)

File: src/confd_gnmi_adapter.py (hunk 2 above)
Comment (micnovak) [reply src/confd_gnmi_adapter.py:292]: Ok, in my opinion this is more readable for me. Do not know why :-)

File: src/confd_gnmi_adapter.py [hunk 6]
```diff
@@ -156,13 +159,18 @@ def is_poll(self):
...
             changes.
             :return:
             """
-            return self.subscription_list.mode == gnmi_pb2.SubscriptionList.STREAM
+            is_monitor = False
+            if self.subscription_list.mode == gnmi_pb2.SubscriptionList.STREAM:
+                for s in  self.subscription_list.subscription:
+                    if s.mode == gnmi_pb2.SubscriptionMode.ON_CHANGE:
+                        is_monitor = True
+            return is_monitor
```
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_adapter.py:173]: I like return to return statement to be as simple as possible. Made my life many times easier.

File: src/confd_gnmi_adapter.py (hunk 4 above)
Comment (micnovak) [reply src/confd_gnmi_adapter.py:289]: OK, I can move it up, just have a feeling it is related only to this part.

File: src/confd_gnmi_adapter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_adapter.py:173]: agree with any

**Review by jomiklos — APPROVED** [review]

**Review by martin-volf — APPROVED** [review]

**Pull request merged**